### PR TITLE
Build support for iPhoneOS Metal Device

### PR DESCRIPTION
**Pull request by adellari**
-Added 'OIDN_METAL_IOS' flag for compiling metal support on ios 
-Modified memory and device initializers for ios compatibility
-Broadened gpu featureset support to include ios A13 chips

### CODE REVIEW

File: core/context.h [hunk 1]
```diff
@@ -71,6 +71,9 @@ OIDN_NAMESPACE_BEGIN
         if (getEnvVarOrDefault("OIDN_DEVICE_METAL", 1))
           OIDN_INIT_STATIC_MODULE(device_metal);
       #endif
+      #if defined(OIDN_METAL_IOS)
+        getEnvVarOrDefault("OIDN_METAL_IOS", 1);
+      #endif
```
Comment (atafra) [inline core/context.h:76]: We shouldn't add a separate environment variable for this. It's still the same Metal device, just the OS is different.

File: core/device_factory.h [hunk 2]
```diff
@@ -29,18 +29,20 @@ OIDN_NAMESPACE_BEGIN
     virtual Ref<Device> newDevice(const int* deviceIDs, const cudaStream_t* streams, int numPairs) = 0;
   };
 
+class MetalDeviceFactoryBase : public DeviceFactory
+{
+public:
+  using DeviceFactory::newDevice;
+  virtual Ref<Device> newDevice(const MTLCommandQueue_id* commandQueues, int numQueues) = 0;
+};
+
```
Comment (atafra) [inline core/device_factory.h:38]: Why was this added?

File: cmake/oidn_bnns.cmake [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 if(NOT IOS)
   set(SDK_VERSION_COMMAND  xcrun -sdk  macosx --show-sdk-version)
-  set(SDK_TARGET 11.0)
+  set(SDK_TARGET 14.0)
```
Comment (atafra) [inline cmake/oidn_bnns.cmake:6]: We cannot raise the SDK target on macOS. This doesn't seem to be relevant for iOS support.

File: include/OpenImageDenoise/config.h.in [hunk 4]
```diff
@@ -76,6 +76,9 @@
 #if !defined(OIDN_DEVICE_METAL)
   #cmakedefine OIDN_DEVICE_METAL
 #endif
+#if !defined(OIDN_METAL_IOS)
+  #cmakedefine OIDN_METAL_IOS
+#endif
```
Comment (atafra) [inline include/OpenImageDenoise/config.h.in:81]: We shouldn't define a macro for this. Please remove.

File: cmake/oidn_metal.cmake [hunk 5]
```diff
@@ -17,6 +17,13 @@ function(metallib_target_add_sources target metallib)
   foreach(def ${METAL_COMPILE_DEFINITIONS})
     list(APPEND compile_defs "-D${def}")
   endforeach()
+  
+  # Select the appropriate SDK based on OIDN_METAL_IOS
+    if(OIDN_METAL_IOS)
+        set(sdk "iphoneos")
+    else()
+        set(sdk "macosx")
+    endif()
```
Comment (atafra) [inline cmake/oidn_metal.cmake:26]: It would be nice to add support for iPadOS too.

File: cmake/oidn_platform.cmake [hunk 6]
```diff
@@ -153,10 +153,17 @@ if((UNIX OR MINGW) AND CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
...
 
+
+#check add a check for the OSX_ARCHITECTURES
 if(APPLE)
   # Make sure code runs on older macOS versions
   if(OIDN_ARCH STREQUAL "ARM64")
-    set(CMAKE_OSX_DEPLOYMENT_TARGET 11.0)
+  #ensure MPS Graph support on ios
+    if(OIDN_METAL_IOS) 
+      set(CMAKE_OSX_DEPLOYMENT_TARGET 14.0)
```
Comment (atafra) [inline cmake/oidn_platform.cmake:163]: Why is the deployment target 14.0 here but for Metal shader compilation only 13.0?

File: devices/metal/metal_engine.h [hunk 7]
```diff
@@ -6,6 +6,12 @@
 
 #include "core/engine.h"
 #include "metal_device.h"
+#include "TargetConditionals.h"
```
Comment (atafra) [inline devices/metal/metal_engine.h:9]: What is this used for?

File: CMakeLists.txt [hunk 8]
```diff
@@ -6,6 +6,9 @@ cmake_minimum_required(VERSION 3.15)
 # Get the library version
 include(cmake/oidn_version.cmake)
 
+# Load the dynamically generated sysroot configuration
+include(sysroot_config.cmake)
```
Comment (atafra) [inline CMakeLists.txt:10]: Do we really need this?

File: sysroot_config.cmake [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+if(OIDN_METAL_IOS)
+  set(SYSROOT_PATH "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk")
+  set(CMAKE_OSX_SYSROOT ${SYSROOT_PATH})
+else()
+  set(SYSROOT_PATH "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk")
+  set(CMAKE_OSX_SYSROOT ${SYSROOT_PATH})
+endif()
+
+file(WRITE ${CMAKE_BINARY_DIR}/sysroot_config.cmake "set(CMAKE_OSX_SYSROOT ${SYSROOT_PATH})")
```
Comment (atafra) [inline sysroot_config.cmake:11]: I don't like that we're hardcoding paths there. I think the sysroot path should be set by the user in CMake.

File: cmake/oidn_bnns.cmake (hunk 3 above)
Comment (adellari) [reply cmake/oidn_bnns.cmake:6]: removing this, uncaught remnant from the debug process.

File: cmake/oidn_metal.cmake (hunk 5 above)
Comment (adellari) [reply cmake/oidn_metal.cmake:26]: Great point, they do share the same sdk. I'll look at building and validating for iPadOS. On that note, the libraries should be the same for xros so that may be worth exploring too.

File: CMakeLists.txt [hunk 10]
```diff
@@ -6,6 +6,9 @@ cmake_minimum_required(VERSION 3.15)
...
+
```
Comment (atafra) [inline CMakeLists.txt:11]: This should be removed. The sysroot is automatically set by CMake when building for iOS with `-DCMAKE_SYSTEM_NAME=iOS`.

File: CMakeLists.txt [hunk 11]
```diff
@@ -33,6 +36,7 @@ if(NOT APPLE)
   option(OIDN_DEVICE_HIP "Enable HIP device." OFF)
 elseif(OIDN_ARCH STREQUAL "ARM64")
   option(OIDN_DEVICE_METAL "Enable Metal device." OFF)
+  option(OIDN_METAL_IOS "Enable iOS Metal device." OFF)
```
Comment (atafra) [inline CMakeLists.txt:39]: We don't need an option for this. If we're building for iOS, the `IOS` CMake variable is set, which we can check instead.

File: devices/metal/CMakeLists.txt [hunk 12]
```diff
@@ -29,13 +29,24 @@ set(OIDN_METAL_SOURCES
...
+else()
+    metallib_target_add_sources(OpenImageDenoise_device_metal metal_kernels
+      metal_kernels.metal
+      INCLUDE_DIRECTORIES
+        ${PROJECT_SOURCE_DIR}
+      COMPILE_OPTIONS
+        -std=metal3.0 -fno-fast-math -mmacosx-version-min=13.0
+    )
+
+endif()
```
Comment (atafra) [inline devices/metal/CMakeLists.txt:49]: It would be nice to merge these two variants and just set the `-mios-version-min`/`-mmacosx-version-min` depending on OS. Also, the iOS target should be 14 instead of 13 because we require 14 anyway due to MPSGraph.

File: CMakeLists.txt (hunk 11 above)
Comment (adellari) [reply CMakeLists.txt:39]: Gotcha, will remove the flag. I added it because I figured manually setting the `SYS_ROOT` and having cmake configure the target might be cumbersome, hence the sysroot_config.cmake file. I'll strip the symbol and have conditionals evaluate `DCMAKE_SYSTEM_NAME` for IOS

File: devices/metal/metal_engine.h (hunk 7 above)
Comment (atafra) [inline devices/metal/metal_engine.h:9]: Should be included as `<TargetConditionals.h>`. I also think that `platform.h` would be a better place for this.

File: CMakeLists.txt (hunk 11 above)
Comment (atafra) [reply CMakeLists.txt:39]: It's sufficient to just evaluate the `IOS` CMake variable, existing code already does that. You don't need to manually set the sysroot if you set `CMAKE_SYSTEM_NAME=iOS`. That's the only CMake that needs to be set by the user to build for iOS.

File: devices/metal/CMakeLists.txt (hunk 12 above)
Comment (atafra) [reply devices/metal/CMakeLists.txt:49]: Actually, we should raise minimum iOS version to 16.0 due to Metal 3.

File: cmake/oidn_platform.cmake (hunk 6 above)
Comment (atafra) [inline cmake/oidn_platform.cmake:163]: We should raise minimum iOS target to 16.0 because Metal 3 is required.

File: core/device_factory.h (hunk 2 above)
Comment (adellari) [reply core/device_factory.h]: Not sure why git diffed this, I reverted the file

File: cmake/oidn_platform.cmake (hunk 6 above)
Comment (adellari) [reply cmake/oidn_platform.cmake]: raised minimum ios target to 16, planning on dealing with deprecation warnings that arise for `MTLResourceOptionCPUCacheModeDefault`

File: devices/metal/metal_engine.h (hunk 7 above)
Comment (adellari) [reply devices/metal/metal_engine.h]: addressed in recent commits

File: CMakeLists.txt (hunk 11 above)
Comment (adellari) [reply CMakeLists.txt]: This should now be addressed. The `SYSROOT` variable is evaluated and `CMAKE_SYSTEM_NAME` is set accordingly, and later evaluated to properly configure gpu device building.

File: sysroot_config.cmake (hunk 9 above)
Comment (adellari) [reply sysroot_config.cmake]: This has now been removed.

File: devices/metal/metal_engine.h [hunk 13]
```diff
@@ -6,6 +6,12 @@
...
+
+#if TARGET_OS_IOS
+#define MANAGED_BUFFERS_UNSUPPORTED
+#define MULTIPLE_DEVICES_UNSUPPORTED
+#endif
```
[the remark this answers is not in the document]
Comment (adellari) [reply devices/metal/metal_engine.h]: `TARGET_OS_IOS` is now directly evaluated instead of defining a new set of macros.

File: cmake/oidn_metal.cmake (hunk 5 above)
Comment (adellari) [reply cmake/oidn_metal.cmake]: Confirmed a working build on iPadOS, added `visionOS` to `CMAKE_SYSTEM_NAME` but I will be unable to confirm a build works as I don't have the hardware. In the meantime, it may be beneficial to set up flags for the iphonesimulator which should make it possible for the library to be used in previews and simulators for ios/ipados.

File: cmake/oidn_platform.cmake (hunk 6 above)
Comment (adellari) [reply cmake/oidn_platform.cmake]: replaced with `MTLResourceCPUCacheModeDefaultCache` in accordance with ios 16's API requirements

File: cmake/oidn_metal.cmake (hunk 5 above)
Comment (adellari) [reply cmake/oidn_metal.cmake]: I think compiling for simulators and investigating xros might be scope creep for this PR, I will investigate in the future.

**Comment by adellari** [comment]
I think the PR should now be up to date with the comments and change requests discussed.

**Comment by atafra** [comment]
Merged to `devel` with some modifications: https://github.com/RenderKit/oidn/commit/983b45b3021a180e955de6e9193b62cfcd38d534
